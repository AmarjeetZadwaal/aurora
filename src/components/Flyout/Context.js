import React from "react";

const { Provider, Consumer } = React.createContext();

export { Provider as FlyoutProvider };
export { Consumer as FlyoutConsumer };
