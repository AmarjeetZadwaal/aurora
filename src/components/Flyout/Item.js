import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import classNames from "classnames";

import colors from "../../theme/colors";
import constants from "../../theme/constants";
import { FlyoutConsumer } from "./Context";
import spacing from "../../theme/spacing";
import typography from "../../theme/typography";

const ItemContainer = styled.div`
  padding: 0 12px;
  height: 36px;
  display: flex;
  align-items: center;
  font-size: ${typography.size.kilo};
  border-radius: ${constants.borderRadius.small};

  &.flyout-item--selected {
    background-color: ${colors.shale};
  }

  &:hover {
    background-color: ${colors.azure.base};
    color: ${colors.white.base};
  }

  &:not(:last-of-type) {
    margin-bottom: ${spacing.slim};
  }
`;

const FlyoutItem = ({ children, index }) => (
  <FlyoutConsumer>
    {({ onItemClick, selectedIndex }) => (
      <ItemContainer
        className={classNames("flyout-item", {
          "flyout-item--selected": index === selectedIndex
        })}
        onClick={() => onItemClick({ index })}
      >
        {children}
      </ItemContainer>
    )}
  </FlyoutConsumer>
);

FlyoutItem.propTypes = {
  children: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired
};

export default FlyoutItem;
