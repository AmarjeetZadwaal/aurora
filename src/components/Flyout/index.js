import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import colors from "../../theme/colors";
import spacing from "../../theme/spacing";
import constants from "../../theme/constants";
import { FlyoutProvider } from "./Context";

const FlyoutContainer = styled.div`
  display: ${props => (props.isOpen ? "block" : "none")};
  flex-direction: column;
  position: absolute;
  top: auto;
  bottom: 90%;
  right: 0;
  width: 205px;
  overflow: hidden;
  background-color: ${colors.white.base};
  padding: ${spacing.cozy};
  border-radius: ${constants.borderRadius.small};
`;

class Flyout extends Component {
  state = {
    onItemClick: () => {}, // eslint-disable-line
    selectedIndex: this.props.selectedIndex // eslint-disable-line
  };

  componentWillMount() {
    this.setState({
      onItemClick: this.onItemClick // eslint-disable-line
    });
  }

  onItemClick = ({ index }) => {
    this.setState({ selectedIndex: index }); // eslint-disable-line
    this.props.onFlyoutOptionSelected({ selectedIndex: index });
  };

  render() {
    const { isOpen, children } = this.props;

    return (
      <FlyoutProvider value={this.state}>
        <FlyoutContainer isOpen={isOpen}>{children}</FlyoutContainer>
      </FlyoutProvider>
    );
  }
}

Flyout.propTypes = {
  children: PropTypes.node.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onFlyoutOptionSelected: PropTypes.func.isRequired,
  selectedIndex: PropTypes.number.isRequired
};

export default Flyout;
