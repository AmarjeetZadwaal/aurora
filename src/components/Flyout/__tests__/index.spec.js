import React from "react";
import { render, Simulate } from "react-testing-library";

import Flyout from "../index";
import FlyoutItem from "../Item";

const mockFn = jest.fn();

const countries = [
  {
    value: "en-US",
    text: "USA (English)",
    label: "United States"
  },
  {
    value: "en-CA",
    text: "Canada (English)",
    label: "Canada"
  },
  {
    value: "fr-CA",
    text: "Canada (Fran√ßais)",
    label: "Canada"
  }
];

describe("<Flyout />", () => {
  it("renders Flyout correctly when its open", () => {
    const { container } = render(
      <Flyout isOpen onFlyoutOptionSelected={mockFn} selectedIndex={0}>
        {countries.map((option, index) => (
          <FlyoutItem index={index} key={option.value}>
            {option.text}
          </FlyoutItem>
        ))}
      </Flyout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders Flyout correctly when its closed", () => {
    const { container } = render(
      <Flyout isOpen={false} onFlyoutOptionSelected={mockFn} selectedIndex={0}>
        {countries.map((option, index) => (
          <FlyoutItem index={index} key={option.value}>
            {option.text}
          </FlyoutItem>
        ))}
      </Flyout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders Flyout correctly when its open", () => {
    const onFlyoutOptionSelected = mockFn;
    const { container } = render(
      <Flyout
        isOpen
        onFlyoutOptionSelected={onFlyoutOptionSelected}
        selectedIndex={0}
      >
        {countries.map((option, index) => (
          <FlyoutItem index={index} key={option.value}>
            {option.text}
          </FlyoutItem>
        ))}
      </Flyout>
    );

    Simulate.click(container.querySelector(".flyout-item"));
    expect(onFlyoutOptionSelected).toHaveBeenCalled();
  });
});
