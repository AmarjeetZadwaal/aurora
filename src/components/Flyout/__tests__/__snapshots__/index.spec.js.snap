// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flyout /> renders Flyout correctly when its closed 1`] = `
<div
  class="sc-bdVaJa hEpNfj"
>
  <div
    class="flyout-item flyout-item--selected sc-bwzfXH jMdqnU"
  >
    USA (English)
  </div>
  <div
    class="flyout-item sc-bwzfXH jMdqnU"
  >
    Canada (English)
  </div>
  <div
    class="flyout-item sc-bwzfXH jMdqnU"
  >
    Canada (Français)
  </div>
</div>
`;

exports[`<Flyout /> renders Flyout correctly when its open 1`] = `
<div
  class="sc-bdVaJa gOdxZs"
>
  <div
    class="flyout-item flyout-item--selected sc-bwzfXH jMdqnU"
  >
    USA (English)
  </div>
  <div
    class="flyout-item sc-bwzfXH jMdqnU"
  >
    Canada (English)
  </div>
  <div
    class="flyout-item sc-bwzfXH jMdqnU"
  >
    Canada (Français)
  </div>
</div>
`;
